src/Microsoft.AspNet.OData.Shared/DeltaOfTStructuralType.cs

```diff
@@ -336,11 +359,75 @@ public void CopyChangedValues(TStructuralType original)
+
+                    if (deltaNestedResource.IsComplexType && !nestedType.IsAssignableFrom(originalType))
+                    {
+                        originalNestedResource = ReAssignComplexDerivedType(original, nestedResourceName, originalNestedResource, nestedType, originalType);
+                    }
+
                     deltaNestedResource.CopyChangedValues(originalNestedResource);
                 }
             }
         }
 
+        private dynamic ReAssignComplexDerivedType(TStructuralType original, string nestedResourceName, dynamic originalNestedResource, Type nestedType, Type originalType)
+        {
+            //As per OASIS discussion, changing a complex type from 1 derived type to another is allowed if both derived type have a common ancestor and the property
+            //is declared in terms of a common ancestor. The logic below checks for a common ancestor. Create a new object of the derived type in delta request.
+            //And copy the common properties.
+
+            Type complexBaseType = nestedType.BaseType;
+            HashSet<Type> lstComplexBaseTypes = new HashSet<Type>();
+
+            while (complexBaseType != null)
+            {
+                lstComplexBaseTypes.Add(complexBaseType);
+                complexBaseType = complexBaseType.BaseType;
+            }
+
+            //Here original type is the type for original (T) resource.
+            //We will keep  going to base types and finally will get the Common Basetype for the derived complex types in to the originalType variable.
+                        
+            bool foundCommonbase = false;
+
+            //The new Original type, means the new complex type (T) which will replace the current complex type.
+            dynamic newOriginalNestedResource = null;
+
+            while (originalType != null)
+            {
+                if (lstComplexBaseTypes.Contains(originalType))
+                {
+                    //Now originalType = common base type of the derived complex types.
+                    //OriginalNested Resource = T(of current Complex type). We are creating newOriginalNestedResource (T - new complex type).
+                    newOriginalNestedResource = Activator.CreateInstance(nestedType);
+
+                    //Here we get all the properties of common base type and get value from original complex type(T) and 
+                    //copy it to the new complex type newOriginalNestedResource(came as a part of Delta) 
+
+                    foreach (PropertyInfo property in originalType.GetProperties())
+                    {
+                        object value = property.GetValue(originalNestedResource);
+                        property.SetValue(newOriginalNestedResource, value);
+                    }
+
+                    foundCommonbase = true;
+
+                    break;
+                }
+
+                originalType = originalType.BaseType;
+            }
+
+            if (foundCommonbase)
```

Comment:
why do you need the foundCommonbase flag at all? why not just put 425 at line 415, before the break;?